File: libsignal/service/src/test/java/org/whispersystems/signalservice/api/storage/SignalStorageCipherTest.java
package org.whispersystems.signalservice.api.storage;

import org.junit.Test;
import org.whispersystems.libsignal.InvalidKeyException;
import org.whispersystems.signalservice.internal.util.Util;

import static org.junit.Assert.assertArrayEquals;

public class SignalStorageCipherTest {

  @Test
  public void symmetry() throws InvalidKeyException {
    StorageItemKey key  = new StorageItemKey(Util.getSecretBytes(32));
    byte[]         data = Util.getSecretBytes(1337);

    byte[] ciphertext = SignalStorageCipher.encrypt(key, data);
    byte[] plaintext  = SignalStorageCipher.decrypt(key, ciphertext);

    assertArrayEquals(data, plaintext);
  }

  @Test(expected = InvalidKeyException.class)
  public void badKeyOnDecrypt() throws InvalidKeyException {
    StorageItemKey key  = new StorageItemKey(Util.getSecretBytes(32));
    byte[]         data = Util.getSecretBytes(1337);

    byte[] badKey = key.serialize().clone();
    badKey[0] += 1;

    byte[] ciphertext = SignalStorageCipher.encrypt(key, data);
    byte[] plaintext  = SignalStorageCipher.decrypt(new StorageItemKey(badKey), ciphertext);
  }
}

File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/storage/SignalStorageCipherTest.kt
package org.whispersystems.signalservice.api.storage

import org.junit.Assert.assertArrayEquals
import org.junit.Test
import org.signal.libsignal.protocol.InvalidKeyException
import org.whispersystems.signalservice.internal.util.Util

class SignalStorageCipherTest {

  @Test
  @Throws(InvalidKeyException::class)
  fun symmetry() {
    val key = StorageItemKey(Util.getSecretBytes(32))
    val data = Util.getSecretBytes(1337)

    val ciphertext = SignalStorageCipher.encrypt(key, data)
    val plaintext = SignalStorageCipher.decrypt(key, ciphertext)

    assertArrayEquals(data, plaintext)
  }

  @Test(expected = InvalidKeyException::class)
  @Throws(InvalidKeyException::class)
  fun badKeyOnDecrypt() {
    val key = StorageItemKey(Util.getSecretBytes(32))
    val data = Util.getSecretBytes(1337)

    val badKey = key.serialize().clone()
    badKey[0] = (badKey[0] + 1).toByte()

    val ciphertext = SignalStorageCipher.encrypt(key, data)
    SignalStorageCipher.decrypt(StorageItemKey(badKey), ciphertext)
  }
}
